File: testsuite/ApplicationCases/SceneEditorApplication/src/application/io/XmlManager.java
//===========================================================================

package application.io;

// JDK basic classes
import java.io.File;
import java.io.IOException;

// XML classes
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.*;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import org.w3c.dom.*;
import org.xml.sax.SAXException;

// VSDK classes
import vsdk.toolkit.common.VSDK;
import vsdk.toolkit.environment.geometry.ParametricCurve;
import vsdk.toolkit.environment.geometry.ParametricBiCubicPatch;
import vsdk.toolkit.io.XmlException;
import vsdk.toolkit.io.geometry.ParametricCurvePersistence;
import vsdk.toolkit.io.geometry.ParametricBiCubicPatchPersistence;

public class XmlManager {
    public static void exportXml(Object object, String outputFilename,
                                 String dtdFilename) throws XmlException {
        try {
            //- 1. Create a new empty Document --------------------------------
            DocumentBuilderFactory factory;
            DocumentBuilder builder;
            Document document;

            factory = DocumentBuilderFactory.newInstance();
            factory.setValidating(false); // Warning: check what it means
            factory.setNamespaceAware(true); // Warning: check what it means
            builder = factory.newDocumentBuilder();
            document = builder.newDocument();

            //- 2. Create an Element from the specified object ----------------
            Element xmlElement;
            xmlElement = null;

            if ( object instanceof ParametricCurve)  {
                ParametricCurve curve = (ParametricCurve) object;
                xmlElement = ParametricCurvePersistence.toElement(curve, 
                                                                  document);
            }
            else if ( object instanceof ParametricBiCubicPatch ) {
                ParametricBiCubicPatch patch = (ParametricBiCubicPatch) object;
                xmlElement = ParametricBiCubicPatchPersistence.toElement(patch,
                                                                     document);
            }

            //- 3. Add Element to Document ------------------------------------
            if ( xmlElement != null ) {
                document.appendChild(xmlElement);
            }

            document.normalizeDocument();

            //- 4. Export Document to File ------------------------------------
            Source source = new DOMSource(document);
            Result result = new StreamResult(new File(outputFilename));

            Transformer xformer;
            xformer = TransformerFactory.newInstance().newTransformer();
            xformer.setOutputProperty(OutputKeys.DOCTYPE_SYSTEM,
                                      dtdFilename);
            // Make resulting XML file more human-readable
            xformer.setOutputProperty(OutputKeys.INDENT, "yes");
            // Warning: what encoding to use?
            xformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");
            xformer.transform(source, result);
        }
        catch ( ParserConfigurationException | XmlException | DOMException | 
                TransformerFactoryConfigurationError | 
                IllegalArgumentException | TransformerException ex ) {
            VSDK.reportMessage(null, VSDK.FATAL_ERROR, "exportXml", "" + ex);
        }
    }

    public static Object importXml(String inputFilename) throws
        XmlException {
        try {
            //- 1. Create a Document from the XML input file ------------------
            DocumentBuilderFactory factory;
            DocumentBuilder builder;
            Document document;

            factory = DocumentBuilderFactory.newInstance();
            factory.setValidating(false); // Warning: check what it means
            factory.setNamespaceAware(true); // Warning: check what it means
            builder = factory.newDocumentBuilder();
            document = builder.parse(new File(inputFilename));

            //- 2. Extract objects from Document ------------------------------
            Node rootNode = document.getFirstChild();
            NodeList nodeList;

            if ( rootNode.getNodeName().equals( 
                 ParametricCurvePersistence.rootName) ) {
                nodeList = document.getElementsByTagName(
                    ParametricCurvePersistence.rootName);
                Node firstNode = nodeList.item(0);
                ParametricCurve curve;
                curve = ParametricCurvePersistence.nodeToParametricCurve(
                    firstNode);
                return curve;
            }
            else if ( rootNode.getNodeName().equals(
                      ParametricBiCubicPatchPersistence.rootName) ) {
                nodeList = document.getElementsByTagName(
                    ParametricBiCubicPatchPersistence.rootName);
                Node firstNode = nodeList.item(0);
                ParametricBiCubicPatch patch;
                patch = ParametricBiCubicPatchPersistence.
                    nodeToParametricBiCubicPatch(firstNode);
                return patch;
            }
            //-----------------------------------------------------------------
        }
        catch ( ParserConfigurationException | SAXException | IOException | 
                XmlException ex ) {
            VSDK.reportMessage(null, VSDK.WARNING, "importXml", "" + ex);
        }
        return null;
    }
}

//===========================================================================
//= EOF                                                                     =
//===========================================================================

File: cpp/testsuite/ApplicationCases/SceneEditorApplication/src/io/XmlManager.cpp
#include "java/io/File.h"
#include "javax/xml/parsers/DocumentBuilder.h"
#include "javax/xml/transform/Transformer.h"
#include "org/w3c/dom/Document.h"
#include "org/w3c/dom/Element.h"
#include "org/w3c/dom/NodeList.h"
#include "vsdk/toolkit/common/logging/Logger.h"
#include "vsdk/toolkit/environment/geometry/curve/ParametricCurve.h"
#include "vsdk/toolkit/environment/geometry/surface/ParametricBiCubicPatch.h"
#include "vsdk/toolkit/io/XmlException.h"
#include "vsdk/toolkit/io/geometry/ParametricBiCubicPatchPersistence.h"
#include "vsdk/toolkit/io/geometry/ParametricCurvePersistence.h"
#include "io/XmlManager.h"

void XmlManager::exportXml(Geometry* object,
                           const java::String& outputFilename,
                           const java::String& dtdFilename)
{
    try {
        //- 1. Create a new empty Document --------------------------------
        org::w3c::dom::Document* document =
            javax::xml::parsers::DocumentBuilder::newDocument();

        //- 2. Create an Element from the specified object ----------------
        org::w3c::dom::Element* xmlElement = nullptr;

        ParametricCurve* curve = dynamic_cast<ParametricCurve*>(object);
        ParametricBiCubicPatch* patch =
            dynamic_cast<ParametricBiCubicPatch*>(object);
        if ( curve != nullptr ) {
            xmlElement = ParametricCurvePersistence::toElement(curve,
                                                              document);
        }
        else if ( patch != nullptr ) {
            xmlElement = ParametricBiCubicPatchPersistence::toElement(patch,
                                                                     document);
        }

        //- 3. Add Element to Document ------------------------------------
        if ( xmlElement != nullptr ) {
            document->appendChild(xmlElement);
        }

        document->normalizeDocument();

        //- 4. Export Document to File ------------------------------------
        javax::xml::transform::Transformer xformer;
        xformer.setOutputProperty(
            javax::xml::transform::Transformer::DOCTYPE_SYSTEM, dtdFilename);
        // Make resulting XML file more human-readable
        xformer.setOutputProperty(
            javax::xml::transform::Transformer::INDENT, "yes");
        // Warning: what encoding to use?
        xformer.setOutputProperty(
            javax::xml::transform::Transformer::ENCODING, "UTF-8");
        if ( !xformer.transform(document, java::File(outputFilename)) ) {
            Logger::reportMessage(java::String(), Logger::FATAL_ERROR,
                "exportXml", java::String("Can not write ") + outputFilename);
        }
        delete document;
    }
    catch ( const XmlException& ex ) {
        Logger::reportMessage(java::String(), Logger::FATAL_ERROR,
            "exportXml", ex.getMessage());
    }
}

Geometry* XmlManager::importXml(const java::String& inputFilename)
{
    //- 1. Create a Document from the XML input file ------------------
    org::w3c::dom::Document* document =
        javax::xml::parsers::DocumentBuilder::parse(java::File(inputFilename));

    if ( document == nullptr ) {
        Logger::reportMessage(java::String(), Logger::WARNING, "importXml",
            java::String("Can not parse ") + inputFilename);
        return nullptr;
    }

    Geometry* result = nullptr;
    try {
        //- 2. Extract objects from Document ------------------------------
        org::w3c::dom::Node* rootNode = document->getDocumentElement();
        org::w3c::dom::NodeList nodeList;

        if ( rootNode->getNodeName().equals(
                 ParametricCurvePersistence::rootName) ) {
            nodeList = document->getElementsByTagName(
                ParametricCurvePersistence::rootName);
            result = ParametricCurvePersistence::nodeToParametricCurve(
                nodeList.item(0));
        }
        else if ( rootNode->getNodeName().equals(
                      ParametricBiCubicPatchPersistence::rootName) ) {
            nodeList = document->getElementsByTagName(
                ParametricBiCubicPatchPersistence::rootName);
            result = ParametricBiCubicPatchPersistence::
                nodeToParametricBiCubicPatch(nodeList.item(0));
        }
        //-----------------------------------------------------------------
    }
    catch ( const XmlException& ex ) {
        Logger::reportMessage(java::String(), Logger::WARNING, "importXml",
                              ex.getMessage());
    }
    delete document;
    return result;
}
